### ISSUE TITLE
read unix @->/tmp/relay-unix periodic error

### ISSUE THREAD

**Issue by estein9825** [paralus/relay#43]
### Expected vs actual behavior

When attempting to connect to certain servers, the kubectl client returns this error:

```
Error from server (InternalError): an error on the server ("read unix @->/tmp/relay-unix-*.core-connector.paralus.iherbpreprod.net: read: connection reset by peer") has prevented the request from succeeding (get pods)
```
or
```
Error from server (unable to proxy kubectl service): ERROR: Unauthenticated access not allowed. Please log in to the portal via browser, or set up API key, for access via the secure kubectl proxy. Error:context deadline exceeded
```

And on the server side I see this issue from the relay-server

```
{"level":"error","ts":"2023-02-23T02:26:21.061Z","caller":"tunnel/server.go:1677","msg":"Relay Server.Server::failed to compete TLS handshake error: remote error: tls: bad certificate  ","stacktrace":"github.com/paralus/relay/pkg/tunnel.(*Server).processDialinTLSState\n\t/build/pkg/tunnel/server.go:1677\ngithub.com/paralus/relay/pkg/tunnel.(*Server).handleDialinConnection\n\t/build/pkg/tunnel/server.go:1794"}
{"level":"error","ts":"2023-02-23T02:26:21.061Z","caller":"tunnel/server.go:1796","msg":"Relay Server.Server.handleDialinConnection::failed to process TLS state error: ailed to compete TLS handshake  ","stacktrace":"github.com/paralus/relay/pkg/tunnel.(*Server).handleDialinConnection\n\t/build/pkg/tunnel/server.go:1796"}
```

Restarting the relay agent of the cluster seems to fix it.

### Steps to reproduce the bug

Uncertain what causes it.

### Are you using the latest version of the project?

You can check your version by running `helm ls|grep '^<deployment-name>'` or using [pctl](https://github.com/paralus/cli), `pctl version`, and provide the output.

```
VERSION: 0.1.0
BUILD: 0.1.0
BUILD-TIME: 1656329967
ARCH: darwin/amd64
```

### What is your environment setup? Please tell us your cloud provider, operating system, and include the output of `kubectl version --output=yaml` and `helm version`. Any other information that you have, eg. logs and custom values, is highly appreciated!

Kubectl
```
clientVersion:
  buildDate: "2022-08-23T17:44:59Z"
  compiler: gc
  gitCommit: a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2
  gitTreeState: clean
  gitVersion: v1.25.0
  goVersion: go1.19
  major: "1"
  minor: "25"
  platform: darwin/amd64
kustomizeVersion: v4.5.7
serverVersion:
  buildDate: "2022-11-29T18:41:42Z"
  compiler: gc
  gitCommit: 52e500d139bdef42fbc4540c357f0565c7867a81
  gitTreeState: clean
  gitVersion: v1.22.16-eks-ffeb93d
  goVersion: go1.16.15
  major: "1"
  minor: 22+
  platform: linux/amd64
```
Helm
```
version.BuildInfo{Version:"v3.10.3", GitCommit:"835b7334cfe2e5e27870ab3ed4135f136eecc704", GitTreeState:"clean", GoVersion:"go1.19.4"}
```

### _(optional)_ If you have ideas on why the bug happens or how it can be solved, please provide it here

The only workaround I've found is to delete the relay agent pod and have it get recreated. But then it happens almost immediately again.

- [x] I've described the bug, included steps to reproduce it, and included my environment setup with all customizations.
- [x] I'm using the latest version of the project.

**Comment by estein9825** [comment]
Apparently on the relay side is an error like this:
```
Agent.Client.paralus-core-relay-agent::dial failed network: tcp  addr: 
17c976c4-dfcd-40d4-a8fa-cb9d31061a94.core-connector.paralus.iherbpreprod.net:443
  err: x509: certificate signed by unknown authority (possibly because 
of "x509: ECDSA verification failure" while trying to verify candidate 
authority certificate "paralus-core-relay")
```
However, it appears to have been swallowed.

**Comment by niravparikh05** [comment]
seems to be related to https://github.com/paralus/paralus/issues/77, will confirm and track one of them if so.

**Comment by estein9825** [comment]
Would be nice if one could perhaps increase the context check return? Just as an FYI, we have about 60 users attempting to hit the problematic cluster regularly.

**Comment by estein9825** [comment]
Another thing we discovered is that the relay server keeps opening up socket connections without ever closing any

```
# k -n paralus exec relay-server-589588bf6c-vz8qf -c relay-server -it -- lsof | grep socket | wc -l
    5825

# k -n paralus exec relay-server-589588bf6c-vz8qf -c relay-server -it -- lsof | grep socket | wc -l
    6020
```

**Comment by niravparikh05** [comment]
Adding some more details which may have caused this issue.

Users are using both the paralus console as well as kubeconfig regularly. And there are around 30 different developers accessing the system regularly with around 10-15 on it at any one time.

All the users are Okta users, none of them are local users. We have configured things to use istio virtualservers, but I saw the same behavior with ingress as well. No envoy/contour. Configured a separate postgres database and audit into postgres and not elasticsearch. We are using fluent-bit. 

NOTE: There are about 30 other clusters in use as well, divided across different projects.

Observed with Paralus version 0.20.

**Comment by niravparikh05** [comment]
Update: 

So the number of imported clusters does not seem to be the primary cause for this issue, even when one cluster is imported after few days the connection between agent and server breaks. web kubectl and even using kubeconfig starts to fail
primary reason causing this is when relay-server tries to establish a proxy unix connection with agent, it fails with below error message in logs
log extracted from relay-server
{"level":"error","ts":"2023-03-13T07:09:15.352Z","caller":"tunnel/server.go:1133","msg":"Relay Server.Server.Listener::io error: error: Put \"https://522b284a-5619-419a-b536-b0c6143173cc.core-connector.ic-oss.dev.rafay-edge.net--cg3k8mqc6p9klgbf1p6g--10.0.0.147:44800\": client not connected  ","stacktrace":"[github.com/paralus/relay/pkg/tunnel.(*Server).proxyUnixConnection](http://github.com/paralus/relay/pkg/tunnel.(*Server).proxyUnixConnection)\n\t/build/pkg/tunnel/server.go:1133"}
from logs there is nothing of importance happening at the agent except for cleanup at this point of time during failure.

**Comment by niravparikh05** [comment]
Update:

problem seems to be with unix sockets created for interaction with upstream k8s api server to handle user requests, for an agent running for almost a week, I see there are multiple sockets created / connected ( close to 5k for relay-agent and 3k for relay-server ) and one in Listening state for both which seems fine but during kubectl while trying to send across messages via socket connection from relay-server to agent the socket conn is reset by agent unix listener for some reason.

1) its possible that all requests are going to a stale conn maintained at the server
2) its possible that agent exhausts the number of sockets / fds it can create and starts to fail

**Comment by niravparikh05** [comment]
@estein9825 we are trying our best to get this in along with helm v0.2.3

**Comment by niravparikh05** [comment]
Update:

looks like current version used in relay of [golang.org/x/net](http://golang.org/x/net) seems to be causing the issue. as a workaround using below pinned version seems to be handling the requests well

replace (
    golang.org/x/net => golang.org/x/net v0.0.0-20210825183410-e898025ed96a
)

**Comment by estein9825** [comment]
So if I fork and generate a simple image with that pin version it should work? Or can I get an unofficial release with that pin?

**Comment by niravparikh05** [comment]
Update on Run 1:

Running load tests for almost an hour by simulating 100 parallel processes below are the stats of the commands executed
total commands fired :: 90000
total commands failed :: 5

the number of fds, sockets opened during this period when randomly checked did not exceed 1k even during peak and was most of the time hovering around 250 or below.

there were only 5 intermittent failures mostly due to fds switch over but it immediately recovered and continued processing requests.

**Comment by estein9825** [comment]
Yep, and I have successfully confirmed that both our production and nonproduction systems are now stable with no issues of connections either through kubeconfig or within the UI itself. Thanks @niravparikh05!

Once a permanent fix is in place this issue can be closed.

**Comment by niravparikh05** [comment]
thanks @estein9825 for validating this, we will proceed with this patch fix for now until we have a permanent solution using the latest version.

**Comment by akshay196** [comment]
I started troubleshooting this. It looks like the problem is with maxConcurrentStreams values set in `golang.org/x/net/http2` package. The relay's dialin client pool check for `CanTakeNewRequest()` function from `net/http2#ClientConn`. The function returns false if number of streams has reached maximum count.
In the commit [e898025ed96a](https://cs.opensource.google/go/x/net/+/e898025ed96aa6d08e98132b8dca210e9e7a0cd2:http2/transport.go;l=642) maxConcurrentStreams was set to 1000. It is now reduced to 100 for some reason.

As the maxConcurrentSteams is local variable we cannot change it. Let me check what we should do if number of requests has reached the maximum.

**Comment by niravparikh05** [comment]
We have been tracking this since last release after upgarding x/net package to v0.9.0 which has fixed this issue. Closing this issue now.

**Comment by pbushuk** [comment]
@niravparikh05 The issue exists after upgrading paralus to v0.2.5 and relay-server to v0.1.6. The count of open sockets increases with every kubectl command until it fails and requires a restart of relay-server.

**Comment by estein9825** [comment]
@niravparikh05 can we reopen this issue? We had to do the `replace golang.org/x/net => golang.org/x/net v0.0.0-20210825183410-e898025ed96a` command for 0.1.6 to not cause the socket error.

**Comment by niravparikh05** [comment]
We have another issue tracking this - [Problem with golang library in Relay server](https://github.com/paralus/paralus/issues/266)
